Graph

robertu94 opened the pull request:
A wide variety of improvement to the graph sections. Including
+ Max Flow algorithms
+ Prim's Minimum Spaning tree
+ A binary heap implementation

robertu94 commented:
@mdclyburn or @ProtractorNinja Could you take a look at this branch before I get some feedback from Dr. Dean?

ProtractorNinja commented:
I'll take a look.

ProtractorNinja commented on algorithms/max-flow/max-flow.tex:
```diff
@@ -0,0 +1,82 @@
+\end{itemize}
+#ifdef hackpackpp
+
+\subsection{Example Contest Problem: Cow-Ex}
```
Why is this called Cow-Ex?

ProtractorNinja commented on algorithms/max-flow/max-flow.tex:
```diff
@@ -0,0 +1,82 @@
+#ifdef hackpackpp
+
+\subsection{Example Contest Problem: Cow-Ex}
+The cows are up to another of the insidious schemes.
```
"the" should be "their".

ProtractorNinja commented on algorithms/max-flow/max-flow.tex:
```diff
@@ -0,0 +1,82 @@
+Each pipe has been rated with a positive integer referring to the amount of milk that the pipe can carry in 1 minute.
+
+The cows want to know how much milk they can transport to the barn at once.
+\subsubsection{Sample Format}
```
Throughout the hackpack, this section is called "Input", "Input Format", or now "Sample Format". For clarity and standardization, the hack pack should only use one.

robertu94 replied:
The exmplar (set) uses Input Format, so I have changed it to that.

ProtractorNinja commented on algorithms/max-flow/max-flow.tex:
```diff
@@ -0,0 +1,82 @@
+#ifdef hackpackpp
+\subsubsection{Lessons Learned}
+\begin{itemize}
+	\item The inner most while loop is a linear solution to the widest most path problem.
```
This should be either clarified or rewritten to make more sense to the average reader.

ProtractorNinja commented on algorithms/max-flow/max-flow.tex:
```diff
@@ -0,0 +1,82 @@
+\section{Max Flows}\index{graph!Max Flows}
+#ifdef hackpackpp
+There are several types of problems that involve finding maximum flow in a graph.
+Many of these problems can easily tackled by making small transformations to the graph.
```
This Max-Flow section does not give a comprehensive overview of problem types, related vocabulary (sources and sinks), or why the following unrelated suggestions are important or useful. It reads more like notes taken during a lecture than a learning resource.

ProtractorNinja commented on algorithms/max-flow/max-flow.tex:
```diff
@@ -0,0 +1,82 @@
+If the flow is limited through the nodes instead of the edges, simply spit every node into a "in" node and an "out" node.
+
+
+\begin{itemize}
```
I think there should be some explanation of why this algorithm is here.

ProtractorNinja commented on algorithms/max-flow/max-flow.tex:
```diff
@@ -0,0 +1,82 @@
+If the graph is undirected, simply insert every edge twice in both directions.
+
+
+If the flow is limited through the nodes instead of the edges, simply spit every node into a "in" node and an "out" node.
```
'a "in"' should be 'an "in"'. "Spit" should be "split".

ProtractorNinja commented on algorithms/max-flow/max-flow.tex:
```diff
@@ -0,0 +1,82 @@
+\subsubsection{Output Format}
+\begin{itemize}
+	\item Line 1: A Positive integer representing the maximum capacity of the network.
+				  This should be 0 if it is not possible to transport the milk
```
Is this extra whitespace intentional? Also, I think the line should end with a period.

ProtractorNinja commented on algorithms/max-flow/problems/cowex/cowex.bats:
```diff
@@ -0,0 +1,7 @@
+#!/usr/bin/env bats
+
+@test "Test in a graph without eddies" {
```
Are eddies explained anywhere?

robertu94 replied:
That is something that I saw in the USACO reference that I looked at.  Essentially they are cycles that artificially boost the max flow by flowing into themselves.  I have not found a graph that does this in practice.  I have asked Dr. Dean about finding an example.

ProtractorNinja replied:
If anything in the hack pack references them, then I think they should be explained somewhere.

ProtractorNinja commented on algorithms/prims-algorithm/prims-algorithm.tex:
```diff
@@ -0,0 +1,63 @@
+\section{Prim's Algorithm}\index{graph!Prim's Algorithm}
+#ifdef hackpackpp
+Prims algorithm is a greedy algorithm that allows for finding a minimum spanning tree/forest in $O(E + \log{V})$ time.
```
Prim's

ProtractorNinja commented on algorithms/prims-algorithm/prims-algorithm.tex:
```diff
@@ -0,0 +1,63 @@
+\section{Prim's Algorithm}\index{graph!Prim's Algorithm}
+#ifdef hackpackpp
+Prims algorithm is a greedy algorithm that allows for finding a minimum spanning tree/forest in $O(E + \log{V})$ time.
+Big picture, Prims algorithm works by doing the following:
```
"Big Picture" makes this line sound unprofessional.

ProtractorNinja commented on algorithms/prims-algorithm/prims-algorithm.tex:
```diff
@@ -0,0 +1,63 @@
+#ifdef hackpackpp
+\subsection{Example Contest Problem: Cow Connection}
+Farmer John has not been very diligent in mowing his farm.
+As a result, the grazing field now resembles a maze with small openings at the intersections.
```
I don't understand what "small openings at the intersections" means. Small clearings? Is it important to mention this?

robertu94 replied:
These are the vertexes,  I will find a better way to say this.

ProtractorNinja commented on algorithms/prims-algorithm/prims-algorithm.tex:
```diff
@@ -0,0 +1,63 @@
+However, Farmer John does not have much extra money to spend on this effort.
+
+He would like to find a set of paths that connect all of the intersections at minimal cost, but is unsure if he can afford it.
+Please Farmer John, find the cost of lighting the maze and the paths to include.
```
This sentence does not very clearly explain what a solution is meant to include. And shouldn't it say something more like "Please help Farmer John to find <something>"?

ProtractorNinja commented on algorithms/prims-algorithm/prims-algorithm.tex:
```diff
@@ -0,0 +1,63 @@
+\subsubsection{Lessons Learned}
+\begin{itemize}
+	\item Typedefs can greatly improve readability of code
+	\item Convince classes can greatly improve readability
```
"Convince classes"?

ProtractorNinja commented on algorithms/prims-algorithm/problems/cowconnection/cowconnection.cpp:
```diff
@@ -1,35 +1,134 @@
 typedef unordered_map<int,int> umii;
-typedef unordered_map< int , umii > umiumii;
-//The graph class uses c++11 extensions; use map to get O(log N) time with c++98
+typedef umii::const_iterator edge_iter;
```
So many "umii"s and "umiumii"s make my eyes water. I think you should change at least one of them to be easier to look at, and so that they're more different. Maybe with fewer i's.

robertu94 replied:
This is just a condensed version.  The full easier to read version is in the section entitled graph.

ProtractorNinja replied:
Something condensed should make every effort to make itself easily understandable, and this naming scheme does quite the opposite. I really think it should be changed.

ProtractorNinja commented on structures/graph/graph.tex:
```diff
@@ -22,25 +22,25 @@ \subsection{Applications}
+	\item Minimum Spanning Trees 
+		\begin{itemize}
+			\item Prim's Algorithm --- for undirected graphs; if run for each component, finds the minimum spanning forest$O(M \log N)$.
+			\item Kruskal's Algorithm --- for graphs; finds the minimum spanning forests in unconnected graphs $(M \log N)$
```
These tabs don't match the rest of the file's indentation method.

ProtractorNinja replied:
Never mind, I think I got confused by GitHub's markup.

ProtractorNinja commented on structures/graph/graph.tex:
```diff
@@ -22,25 +22,25 @@ \subsection{Applications}
+            \item Breath First Search --- Graphs with equal weight edges $O(N+M)$
+            \item Dijkstra's Algorithm --- Graphs with non-negative edges weights $O(N \log N)$.
+            \item Bellman Ford --- Graphs with some negative edges $O(N^3)$
+            \item Floyd Warshall --- Grahps with negative edges but not cycles $O(NM)$.
```
Grahps

ProtractorNinja commented on structures/graph/graph.tex:
```diff
@@ -22,25 +22,25 @@ \subsection{Applications}
-            \item Depth First Search - Find (Strongly) Connected Components $O(N+M)$
-            \item Depth First Search - Find a path from i to j $O(N+M)$
+            \item Breath First Search --- Graphs with equal weight edges $O(N+M)$
+            \item Dijkstra's Algorithm --- Graphs with non-negative edges weights $O(N \log N)$.
```
edges weights -> edge weights

ProtractorNinja commented on structures/heap/heap.tex:
```diff
@@ -1 +1,25 @@
-\section{Heaps}
+\section{Heaps}\index{heap}\index{priority\_queue}
```
"Heaps" are plural, but the other data structures are singular.

ProtractorNinja commented on structures/heap/heap.tex:
```diff
@@ -1 +1,25 @@
-\section{Heaps}
+\section{Heaps}\index{heap}\index{priority\_queue}
+Heaps are very useful data structures that support at least the following operations:
+
```
This is not a very comprehensive explanation of heaps, lacking more time complexity information, what they are particularly good at, or other useful details. That said, it is better than nothing.

ProtractorNinja commented on structures/heap/heap.tex:
```diff
@@ -1 +1,25 @@
+Many other implementations also implement a decrease key operation and delete operation.
+
+The standard library provides a two sets of functions that provide this functionality.
+The first is the priority\_queue data structure found in the queue header.
```
Do we not have inline code tags for priority_queue or make_heap?

ProtractorNinja commented:
As you can see, I made a bunch of remarks on the code.

Generally speaking, a lot of the new (and old, actually) text is kind of a janky mashup between super-concise, uselessly verbose, or awkwardly composed writing that makes the text feel as if it was not given a lot of thought.

mdclyburn commented on algorithms/max-flow/max-flow.tex:
```diff
@@ -0,0 +1,82 @@
+Then find the flow from the super source to the super sink.
+
+
+If the graph is undirected, simply insert every edge twice in both directions.
```
Should this really be the case if the graph is _undirected_, or should it be if the graph is _directed_?

robertu94 replied:
The underlying implementation is directed, so to mimic directed functionality, you must double the edges.

ProtractorNinja replied:
That fact isn't immediately clear (it confused both me and Marshall), so it ought to be clarified somehow.

mdclyburn commented on algorithms/max-flow/max-flow.tex:
```diff
@@ -0,0 +1,82 @@
+\end{itemize}
+
+\subsubsection{Sample Input}
+\acmlisting[label=cowex sample input, caption=cowex sample input]{./algorithms/max-flow/problems/cowex/cowex.in}
```
The caption and label should be capitalized for consistency with the rest of the document.

mdclyburn commented on algorithms/max-flow/max-flow.tex:
```diff
@@ -0,0 +1,82 @@
+		This should be 0 if it is not possible to transport the milk.
+\end{itemize}
+\subsubsection{Sample Output}
+\acmlisting[label=cowex sample output, caption=cowex sample output]{./algorithms/max-flow/problems/cowex/cowex.out}
```
Same capitalization nitpick.

mdclyburn commented on algorithms/max-flow/max-flow.tex:
```diff
@@ -0,0 +1,82 @@
+\subsection{Greedy Max Flow Algorithm}
+#endif
+
+\acmlisting[label=cowex sample output, caption=cowex sample output]{./algorithms/max-flow/problems/cowex/cowex.cpp}
```
Same capitalization nitpick.

mdclyburn commented on algorithms/max-flow/problems/cowex/Makefile:
```diff
@@ -0,0 +1,9 @@
+CXX=clang++
```
Should we really be using `clang++` here? It should be fine, but the rest of the problems use contest standards.

mdclyburn commented on algorithms/prims-algorithm/prims-algorithm.tex:
```diff
@@ -0,0 +1,63 @@
+Farmer John has not been very diligent in mowing his farm.
+As a result, the grazing field now resembles a maze with small openings at the intersections.
+
+At night, this maze is unlit, and is very dangerous.
```
The second comma needs to die. 'is very dangerous' is not an independent clause (haha). Or you can make it '_it_ is very dangerous'.

mdclyburn commented on algorithms/prims-algorithm/prims-algorithm.tex:
```diff
@@ -0,0 +1,63 @@
+\end{itemize}
+
+\subsubsection{Sample Input}
+\acmlisting[label=cowconnection sample input, caption=cowconnection sample input]{./algorithms/prims-algorithm/problems/cowconnection/cowconnection.in}
```
Capitalize the label and caption for consistency.

mdclyburn commented on algorithms/prims-algorithm/prims-algorithm.tex:
```diff
@@ -0,0 +1,63 @@
+	\item Line 2..$(N+1)$: Two integers, the sets of intersections that are endpoints in the paths Farmer John should include. These integers should be in order from least to greatest.
+\end{itemize}
+\subsubsection{Sample Output}
+\acmlisting[label=cowconnection sample output, caption=cowconnection sample output]{./algorithms/prims-algorithm/problems/cowconnection/cowconnection.out}
```
Same nitpick as above.

mdclyburn commented on algorithms/prims-algorithm/prims-algorithm.tex:
```diff
@@ -0,0 +1,63 @@
+#ifdef hackpack
+\subsection{Prim's Algorithm}
+#endif
+\acmlisting[label=cowconnection example solution, caption=cowconnection example solution]{./algorithms/prims-algorithm/problems/cowconnection/cowconnection.cpp}
```
Same nitpick as above.

bcdean commented:
In heap description, "Remove the smallest element" should be generalized to indicate that you can also build a heap where you want to remove the largest element (in fact this is what the STL priority queue does by default).
Running times not quite right: Dijkstra should be O(M log N), Floyd Warshall should be O(N^3), Dynamic Programming should be O(M), Bellman-Ford should be O(MN).  Prim in 2.8 is listed as runtime O(E+log V) which is incorrect; the usual running time is equal to that of Dijkstra, which is O(M log N) (or O(E log V)) when used with a binary heap.  The "Lessons Learned" note about wiest most path taking O(V) time is incorrect -- it is O(E log V).  Running time O(FE) quoted for max flow is true regardless of whether we use the widest path or not; with widest path, it is actually O(EV log V log F).
The part in the description about "reduce the cost of the nodes on the widest path by its width" and "create reverse paths..." I found a bit vague; this could be written more clearly so it's easier to understand.

robertu94 commented:
@bcdean I think I have addressed the issues that you brought up in your last comment.  I also changed all of the running time to be in terms of E,V,and F because that seems more standard.

bcdean commented:
Looks much better.  -- Brian

From: Robert Underwood [mailto:notifications@github.com]
Sent: Wednesday, November 11, 2015 1:04 PM
To: clemsonacm/hackpack
Cc: Brian Dean
Subject: Re: [hackpack] Graph (#35)


@bcdean<https://github.com/bcdean> I think I have addressed the issues that you brought up in your last comment. I also changed all of the running time to be in terms of E,V,and F because that seems more standard.

—
Reply to this email directly or view it on GitHub<https://github.com/clemsonacm/hackpack/pull/35#issuecomment-155863140>.
